### PR TITLE
feat(oauth):support oauth way to create project

### PR DESCRIPTION
**Pull request by wzqnls**
**What this PR does / why we need it**:

Add your description

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to [#576](https://github.com/caicloud/service/issues/576)

**Special notes for your reviewer**:

/cc @zhujian7 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/api/v1/handler/oauth.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/caicloud/cyclone/cmd/worker/options"
```
Comment (zhujian7) [inline pkg/api/v1/handler/oauth.go]: Please format import

File: pkg/api/v1/handler/oauth.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+// then return url to frontend
+func GetAuthCodeURL(ctx context.Context, scm string) (string, error) {
+	switch scm {
+	case "gitlab":
+		url := gitlabConf.AuthCodeURL(fakeState)
+		log.Infof("new redirect url for gitlab oauth2: %s", url)
+		return url, nil
+	case "github":
+		//TODO: support github oauth2
+		return "", nil
```
Comment (zhujian7) [inline pkg/api/v1/handler/oauth.go]: I think to to return Not Implemented error is better than nil for github, and add default case is needed.

File: pkg/api/v1/handler/oauth.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+			consoleWebServer, code, state, oauthToken.AccessToken, userName, server,
+		)
+
+		httpCtx := service.HTTPContextFrom(ctx)
+		w := httpCtx.ResponseWriter()
+		r := httpCtx.Request()
+		http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
+
+		return nil
+	case "github":
```
Comment (zhujian7) [inline pkg/api/v1/handler/oauth.go]: Same as above

File: pkg/api/v1/handler/oauth.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	return "", nil
+}
+
+// GetToken use code to change token and redirect to frontend
+func GetToken(ctx context.Context, scm string, code string, state string) error {
+	if code == "" || state != fakeState {
+		return fmt.Errorf("code is nil or state is not right: code %s, state %s", code, state)
+	}
+	switch scm {
+	case "gitlab":
```
Comment (zhujian7) [inline pkg/api/v1/handler/oauth.go]: how about to make "gitlab" "github" as Scm type?

File: pkg/api/v1/handler/oauth.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+		return nil
+	case "github":
+		//TODO: support github oauth2
+		return nil
+	}
+	return nil
+}
+
+//TODO: support gitlab API V4 version
```
[the remark this answers is not in the document]
Comment (wzqnls) [reply pkg/api/v1/handler/oauth.go]: 大神说啥就是啥 这两天可以思考下设计方案。本来想写个English，怕写成了Chinlish。

**Comment by supereagle** [comment]
Not needed any more.

/close

**Pull request closed without merging**